File: apps/coordinator-api/app/schemas/vault.py
"""
schemas/vault.py — ShardLock Coordinator API
=============================================
Pydantic schemas for vault management endpoints.

Mirrors the pattern in schemas/auth.py — request/response pairs
per endpoint, all inheriting from BaseModel.

Per §1.3 Standard API Response Contract:
  - Success: success=True, data object, message
  - Error: handled via HTTPException (success=False, error code, message)
"""

from datetime import datetime
from typing import Optional, List
from uuid import UUID

from pydantic import BaseModel, Field


# ── Create Vault ──────────────────────────────────────────────────────────────

class VaultCreateRequest(BaseModel):
    site_name       : str  = Field(..., min_length=1, max_length=255)
    username        : str  = Field(..., min_length=1, max_length=255)
    plaintext_password: str = Field(..., min_length=1)
    master_password : str  = Field(..., min_length=1)
    label           : Optional[str] = Field(None, max_length=255)


class VaultCreateResponse(BaseModel):
    success         : bool
    message         : str
    vault_id        : str


# ── List Vaults ───────────────────────────────────────────────────────────────

class VaultEntryMeta(BaseModel):
    """Single vault entry metadata — no passwords, no encrypted data."""
    id          : UUID
    site_name   : str
    username    : str
    label       : Optional[str]
    created_at  : datetime
    updated_at  : datetime

    class Config:
        from_attributes = True


class VaultListResponse(BaseModel):
    success     : bool
    data        : List[VaultEntryMeta]
    total       : int
    page        : int
    page_size   : int


# ── Get Single Vault (triggers reconstruction) ────────────────────────────────

class VaultRetrieveResponse(BaseModel):
    success             : bool
    vault_id            : str
    site_name           : str
    username            : str
    label               : Optional[str]
    plaintext_password  : str   # decrypted, returned to client — never stored


# ── Delete Vault ──────────────────────────────────────────────────────────────

class VaultDeleteResponse(BaseModel):
    success : bool
    message : str

File: apps/frontend/src/api/vault.ts
import { apiClient } from './client'

// ── Types ─────────────────────────────────────────────────────────────────────

export interface VaultEntryMeta {
  id: string
  site_name: string
  username: string
  label: string | null
  created_at: string
  updated_at: string
}

export interface VaultListResponse {
  success: boolean
  data: VaultEntryMeta[]
  total: number
  page: number
  page_size: number
}

export interface VaultCreateRequest {
  site_name: string
  username: string
  plaintext_password: string
  master_password: string
  label?: string
}

export interface VaultCreateResponse {
  success: boolean
  message: string
  vault_id: string
}

export interface VaultRetrieveResponse {
  success: boolean
  vault_id: string
  site_name: string
  username: string
  label: string | null
  plaintext_password: string
}

export interface VaultDeleteResponse {
  success: boolean
  message: string
}

// ── API calls ─────────────────────────────────────────────────────────────────

export const vaultApi = {
  list: async (page = 1, page_size = 20): Promise<VaultListResponse> => {
    const { data } = await apiClient.get<VaultListResponse>('/vault', {
      params: { page, page_size },
    })
    return data
  },

  create: async (payload: VaultCreateRequest): Promise<VaultCreateResponse> => {
    const { data } = await apiClient.post<VaultCreateResponse>('/vault', payload)
    return data
  },

  retrieve: async (
    vault_id: string,
    master_password: string
  ): Promise<VaultRetrieveResponse> => {
    const { data } = await apiClient.get<VaultRetrieveResponse>(
      `/vault/${vault_id}`,
      { params: { master_password } }
    )
    return data
  },

  delete: async (vault_id: string): Promise<VaultDeleteResponse> => {
    const { data } = await apiClient.delete<VaultDeleteResponse>(
      `/vault/${vault_id}`
    )
    return data
  },
}